homeassistant/components/enphase_envoy/sensor.py

```diff
@@ -51,33 +58,71 @@
             cv.ensure_list, [vol.In(list(SENSORS))]
         ),
         vol.Optional(CONF_NAME, default=""): cv.string,
+        vol.Optional(CONF_SCAN_INTERVAL, default=60): cv.time_period,
     }
 )
 
 
-async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+async def async_setup_platform(
+    homeassistant, config, async_add_entities, discovery_info=None
+):
     """Set up the Enphase Envoy sensor."""
     ip_address = config[CONF_IP_ADDRESS]
     monitored_conditions = config[CONF_MONITORED_CONDITIONS]
     name = config[CONF_NAME]
     username = config[CONF_USERNAME]
     password = config[CONF_PASSWORD]
+    scan_interval = config[CONF_SCAN_INTERVAL]
 
     envoy_reader = EnvoyReader(ip_address, username, password)
 
+    async def async_update_data():
+        """Fetch data from API endpoint."""
+        async with async_timeout.timeout(30):
+            i = 0
+            while i < 3:
+                try:
+                    data = await envoy_reader.update()
+                except httpcore.ProtocolError as err:
+                    _LOGGER.error("Error communicating with API: %s", err)
+                    break
+                except httpcore.ConnectTimeout as err:
+                    _LOGGER.error("Timeout error with API: %s", err)
+                    break
+
+                _LOGGER.debug("Retrieved data from API: %s", data)
+
+                if "can't handle event type ConnectionClosed" in str(
+                    data.get("inverters_production")
+                ):
+                    _LOGGER.debug("Retry polling Envoy.  Previous attempt failed.")
+                    i += 1
+                else:
+                    _LOGGER.debug("Returning API data.")
+                    return data
+
+            # After 3 communication errors log the message and wait
+            # for the next polling cycle
+            _LOGGER.warning(
+                "Communication error with Enphase Envoy.  Will retrieve data on the next poll."
+            )
+
+    coordinator = DataUpdateCoordinator(
+        homeassistant,
+        _LOGGER,
+        name="sensor",
+        update_method=async_update_data,
+        update_interval=scan_interval,
+    )
+
+    await coordinator.async_refresh()
+
     entities = []
     # Iterate through the list of sensors
     for condition in monitored_conditions:
         if condition == "inverters":
-            try:
-                inverters = await envoy_reader.inverters_production()
-            except requests.exceptions.HTTPError:
-                _LOGGER.warning(
-                    "Authentication for Inverter data failed during setup: %s",
-                    ip_address,
-                )
-                continue
-
+            inverters = coordinator.data.get("inverters_production")
```

Comment:
Added a check if the data is `NoneType` than `raise PlatformNotReady`

Reply:
- Removed `should_poll`
- Removed `available`
- Removed `async_added_to_hass`
- Removed `async_update` and moved the code into `state`
- In `state` it wasn't a copy and paste from `async_update`, but mostly the same code exists. Tried to reduce the paths and catch the same errors. I moved the update for the inverter `last_reported` attribute into the `device_state_attributes()` method

I noticed now the sensors have values on the first poll rather than waiting for the second poll!